### PR TITLE
nixos/mautrix-signal: add module

### PR DESCRIPTION
**Pull request by niklaskorz**
## Description of changes

Adds a NixOS service module for [mautrix-signal](https://github.com/mautrix/signal/), based on the existing module for mautrix-whatsapp.
This depends on https://github.com/NixOS/nixpkgs/pull/277367 as the service configuration was only tested with the new Go version of mautrix-signal, I haven't tested if it is compatible with the now broken Python version.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Swarsel** [comment]
What is the status on this? I would love to see this module added :)

**Comment by niklaskorz** [comment]
> What is the status on this? I would love to see this module added :)

Sorry, it kind of went under the radar for me. 😅 I'll apply the feedback this weekend and will then look for a second reviewer.

File: nixos/modules/services/matrix/mautrix-signal.nix
Comment (symphorien) [reply nixos/modules/services/matrix/mautrix-signal.nix:131]: I'm not sure mkDefaults works, when trying do deploy this I had a conflict over services.mautrix-signal.settings.appservice.database

File: nixos/modules/services/matrix/mautrix-signal.nix
Comment (symphorien) [inline nixos/modules/services/matrix/mautrix-signal.nix:100]: I cannot find any documentation on this environment variable, can you add a pointer to its usage?

**Review by symphorien — COMMENTED** [review]
I could successfully deploy this bridge with this moduie. the path to the registration file is somewhat nixos specific I think so it would be nice to indicate in the documentation of one of the options that you need this additional configuration:
```
  services.matrix-synapse = {                                                            
    settings.app_service_config_files = [                                                                              
    "/var/lib/mautrix-signal/signal-registration.yaml"                                   
  ];                                                                                     
  # ...                                                                                  
};                                                                                       
systemd.services.matrix-synapse.serviceConfig.SupplementaryGroups = [ "mautrix-signal" ];
};                                                                                       
```

Or include this configuration by default ?

**Comment by symphorien** [comment]
@niklaskorz since 24.05 is not far away this would be a good occasion to push this through the finish line

File: nixos/modules/services/matrix/mautrix-signal.nix
Comment (niklaskorz) [reply nixos/modules/services/matrix/mautrix-signal.nix:100]: I double checked myself and I think you are correct. The [docs](https://docs.mau.fi/bridges/general/double-puppeting.html) only reference the YAML way of configuring this, which is also what I'm using on my own server. I originally copied this over from the mautrix-whatsapp module, but I ran into the same issue there and had to put the secret into the config attribute set instead. Maybe @frederictobiasc (the maintainer of the mautrix-whatsapp module) can shine some light on this?

Comment (niklaskorz) [reply nixos/modules/services/matrix/mautrix-signal.nix:100]: To be more precise, I think this environment variable has been removed in both mautrix-signal and mautrix-whatsapp when they were rewritten in Go.

Comment (niklaskorz) [reply nixos/modules/services/matrix/mautrix-signal.nix:100]: @Ma27 Are you using double puppeting with your Signal bridge? If so, how do you configure it?

Comment (niklaskorz) [reply nixos/modules/services/matrix/mautrix-signal.nix:100]: > To be more precise, I think this environment variable has been removed in both mautrix-signal and mautrix-whatsapp when they were rewritten in Go.

Never mind that, sorry, it's been awhile since setting this up. :) The service file is using envsubst to generate the file configuration, we only have to adjust the name and format to fit the new secret "map", which is needed because the secrets are now configured per domain. I'll look into it.

File: nixos/modules/services/matrix/mautrix-signal.nix
Comment (niklaskorz) [reply nixos/modules/services/matrix/mautrix-signal.nix]: @symphorien I changed the settings option to apply updates recursively and now only set the homeserver config here using mkDefaults. Works fine in my setup, does this work on your end too?

**Comment by niklaskorz** [comment]
> I could successfully deploy this bridge with this moduie. the path to the registration file is somewhat nixos specific I think so it would be nice to indicate in the documentation of one of the options that you need this additional configuration:
> 
> ```
>   services.matrix-synapse = {                                                            
>     settings.app_service_config_files = [                                                                              
>     "/var/lib/mautrix-signal/signal-registration.yaml"                                   
>   ];                                                                                     
>   # ...                                                                                  
> };                                                                                       
> systemd.services.matrix-synapse.serviceConfig.SupplementaryGroups = [ "mautrix-signal" ];
> };                                                                                       
> ```
> 
> Or include this configuration by default ?

I added an option `registerToSynapse` which is enabled by default if the `matrix-synapse` service is enabled. Can you test it? I'm personally not using Synapse (I'm on Conduit).

File: nixos/modules/services/matrix/mautrix-signal.nix
Comment (symphorien) [inline nixos/modules/services/matrix/mautrix-signal.nix:104]:
```suggestion
        File containing environment variables to be passed to the mautrix-signal service.
        If an environment variable `MAUTRIX_SIGNAL_BRIDGE_LOGIN_SHARED_SECRET` is set, then its value will be used in the configuration file for the option `login_shared_secret_map` without leaking it to the store. See [here](https://github.com/mautrix/signal/blob/main/example-config.yaml) for the documentation of `login_shared_secret_map`.
```

Comment (niklaskorz) [reply nixos/modules/services/matrix/mautrix-signal.nix]: Applied your suggestion and added a hint that the `homeserver.domain` will be used as key inside the shared secret map in this case. Thanks!

**Review by symphorien — APPROVED** [review]
the automated synapse configuration works fine. I'll merge after other people have had time to provide further feedback.

**Review by cyclic-pentane — APPROVED** [review]

**Pull request merged**
